### PR TITLE
Only ensure the ThreadPool is active when queuing tasks

### PR DESCRIPTION
**Pull request by benaadams**
Rather than double requesting threads on both `Enqueue` and `Dispatch`, only ensure there is >= 1  requests outstanding at enqueue, and let dispatch request up to `ProcessorCount` threads; rather than having `ProcessorCount` as the cap in both places.

Idea from @kouvel's comment in https://github.com/dotnet/runtime/pull/35330#issue-407812635

> It also doesn't help that the thread pool requests many more threads than necessary for the number of work items enqueued

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs [hunk 1]
```diff
@@ -424,9 +424,28 @@ private ThreadPoolWorkQueueThreadLocals CreateThreadLocals()
             return ThreadPoolWorkQueueThreadLocals.threadLocals = new ThreadPoolWorkQueueThreadLocals(this);
         }
 
+        internal void EnsureThreadPoolActive()
+        {
+            // Used for queuing.
+
+            // If we have not yet requested a thread, then request a new thread so the ThreadPool is active.
+            int count = numOutstandingThreadRequests;
+            while (count < 1)
```
Comment (kouvel) [inline src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:433]: I believe this speculative check is insufficient to guarantee that the thread pool would be active on an enqueue.

I'm already looking into this problem and I think I have a decent approach that would be correct along with compensating for the extra overhead of a non-speculative check here. I'm currently looking at doing that along with my changes to switch the coreclr runtime to use the portable thread pool (WIP changes in my branch `ThreadPool` at the moment), as this type of change would involve a fair bit of testing that I would need to do with those changes anyway. If for some reason the portable thread pool change cannot be merged for 5.0 then there are several portions of that PR that I would plan to submit separately including a fix for this issue. Let me try a fix for it and see where it goes. If you have a scenario in mind where this type of change would be impactful I would greatly appreciate it if you could help to verify those.

Comment (kouvel) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:433]: Edited above

Comment (benaadams) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:433]: 👍

Comment (davidfowl) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:433]: > If for some reason the portable thread pool change cannot be merged for 5.0 then there are several portions of that branch that I would plan to submit separately including a fix for this issue. Let me try a fix for it and see where it goes. If you have a scenario in mind where this type of change would be impactful I would greatly appreciate it if you could help to verify those, though that branch is not ready for verification yet. I'll ping on on the PR once I have it up.

No negative thoughts, we'll get it in for 5.0 😄

Comment (davidfowl) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:433]: @kouvel where's your branch?

Comment (kouvel) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:433]: It's this one: https://github.com/kouvel/runtime/tree/ThreadPoolRequestFewerThreads
I will likely update the various commits with force-push so the commits may change over time (to make it easier to review) until I put up a PR.

Comment (kouvel) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:433]: Eh that's the wrong one

Comment (kouvel) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:433]: Updated inline above

Comment (kouvel) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:433]: It doesn't have a fix for this issue yet, I have a prototype that I'll be adding to that branch

Comment (davidfowl) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:433]: Awesome! I'll be stalking these changes 😄 .

**Pull request closed without merging**